### ISSUE TITLE
Illegal instruction in version 1.0.12

### ISSUE THREAD

**Issue by CHERTS** [akopytov/sysbench#212]
I use sysbench 1.0.12 on Debian  8.10 (jessie) in Virtuozzo Contaner.

I run sysbench with a minimum of options and got an error:

$ sysbench cpu help
Illegal instruction

$ sysbench cpu --cpu-max-prime=20000 run
Illegal instruction

Linux kernel:
$ uname -a
Linux xxxx.ru 3.16.0-042stab120.18 #1 SMP Tue Oct 17 12:48:22 MSK 2017 x86_64 GNU/Linux

What can I do to start sysbench?

**Comment by akopytov** [comment]
The most likely reasons is that the sysbench binary you are using was compiled with optimizations for a newer CPU. For example, binary with `-march=broadwell` will trigger an illegal instruction when executed on a Haswell CPU. Can you tell me more about how this sysbench binary was installed?

**Comment by CHERTS** [comment]
I installed the latest version of sysbench according to the instructions:

> curl -s https://packagecloud.io/install/repositories/akopytov/sysbench/script.deb.sh | sudo bash
> sudo apt -y install sysbench

Here is information about the binary file:

$ file /usr/bin/sysbench
/usr/bin/sysbench: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, for GNU/Linux 2.6.32, BuildID[sha1]=19b7e10c9a6fb3844621da4e07197697a94a465b, stripped

$ readelf -h /usr/bin/sysbench
ELF Header:
  Magic:   7f 45 4c 46 02 01 01 00 00 00 00 00 00 00 00 00
  Class:                             ELF64
  Data:                              2's complement, little endian
  Version:                           1 (current)
  OS/ABI:                            UNIX - System V
  ABI Version:                       0
  Type:                              EXEC (Executable file)
  Machine:                           Advanced Micro Devices X86-64
  Version:                           0x1
  Entry point address:               0x40917a
  Start of program headers:          64 (bytes into file)
  Start of section headers:          662792 (bytes into file)
  Flags:                             0x0
  Size of this header:               64 (bytes)
  Size of program headers:           56 (bytes)
  Number of program headers:         9
  Size of section headers:           64 (bytes)
  Number of section headers:         28
  Section header string table index: 27

$ strings -a  /usr/bin/sysbench | grep -i gcc
libgcc_s.so.1
GCC_3.3
GCC_3.0
gcconsts

$ ldd /usr/bin/sysbench
        linux-vdso.so.1 (0x00007fff0e5cb000)
        libmysqlclient.so.18 => /usr/lib/x86_64-linux-gnu/libmysqlclient.so.18 (0x00007fe503f66000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007fe503d49000)
        libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007fe503b2e000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007fe50392a000)
        libaio.so.1 => /lib/x86_64-linux-gnu/libaio.so.1 (0x00007fe503728000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fe503427000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007fe503211000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fe502e66000)
        libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007fe502b5b000)
        /lib64/ld-linux-x86-64.so.2 (0x00007fe5044a9000)

**Comment by akopytov** [comment]
Thanks. Can you also show me the output from `cat /proc/cpuinfo` or `lscpu`?

**Comment by akopytov** [comment]
Actually, the `ldd` output looks suspicious. Packages from packagecloud.io are linked with both MySQL and PostgreSQL client libraries, but I don't see `libpq.so` in the dependencies. Either the output is truncated, or that sysbench binary was actually installed from elsewhere.

**Comment by CHERTS** [comment]
ldd output not truncated and sysbench binary was installed from packagecloud.io

I tested this on 3 servers using the md5 checksum
$ md5sum /usr/bin/sysbench
438e850ddb8f0b36ff41f254fd046905  /usr/bin/sysbench

$ cat /proc/cpuinfo
processor       : 0
vendor_id       : GenuineIntel
cpu family      : 6
model           : 44
model name      : Intel(R) Xeon(R) CPU           E5645  @ 2.40GHz
stepping        : 2
microcode       : 21
cpu MHz         : 2401.000
cache size      : 12288 KB
physical id     : 0
siblings        : 12
core id         : 0
cpu cores       : 6
apicid          : 0
initial apicid  : 0
fpu             : yes
fpu_exception   : yes
cpuid level     : 11
wp              : yes
flags           : fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts rep_good xtopology nonstop_tsc aperfmperf pni dtes64 monitor ds_cpl vmx smx est tm2 ssse3 cx16 xtpr pdcm pcid dca sse4_1 sse4_2 popcnt lahf_lm ida arat epb dtherm tpr_shadow vnmi flexpriority ept vpid
bogomips        : 4800.41
clflush size    : 64
cache_alignment : 64
address sizes   : 40 bits physical, 48 bits virtual
power management:

processor       : 1
vendor_id       : GenuineIntel
cpu family      : 6
model           : 44
model name      : Intel(R) Xeon(R) CPU           E5645  @ 2.40GHz
stepping        : 2
microcode       : 21
cpu MHz         : 2401.000
cache size      : 12288 KB
physical id     : 0
siblings        : 12
core id         : 1
cpu cores       : 6
apicid          : 2
initial apicid  : 2
fpu             : yes
fpu_exception   : yes
cpuid level     : 11
wp              : yes
flags           : fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts rep_good xtopology nonstop_tsc aperfmperf pni dtes64 monitor ds_cpl vmx smx est tm2 ssse3 cx16 xtpr pdcm pcid dca sse4_1 sse4_2 popcnt lahf_lm ida arat epb dtherm tpr_shadow vnmi flexpriority ept vpid
bogomips        : 4800.41
clflush size    : 64
cache_alignment : 64
address sizes   : 40 bits physical, 48 bits virtual
power management:

$ lscpu
lscpu: failed to determine number of CPUs: /sys/devices/system/cpu/possible: No such file or directory

lscpu gives an error because this server is virtual (OpenVZ/Virtuozzo virtualization)

I build sysbench from deb source:
$ apt-get source sysbench
$ apt-get build-dep sysbench
$ cd sysbench-1.0.12/
$ dpkg-buildpackage -b

and it works fine:

$ ./debian/sysbench/usr/bin/sysbench cpu help
sysbench 1.0.12 (using bundled LuaJIT 2.1.0-beta2)

cpu options:
  --cpu-max-prime=N upper limit for primes generator [10000]

**Comment by akopytov** [comment]
OK, E5645 appears to be Westmere, which is older than everything I currently have access to. I may need to revisit my Travis CI scripts. But the problem is clear now, thank you.

**Comment by CHERTS** [comment]
I found another server with a similar problem, it's a baremetal server and it also issues an Illegal instruction.

# lsb_release -a
Description:    Debian GNU/Linux 8.10 (jessie)
Release:        8.10
Codename:       jessie

Here is the output of lscpu from this server:
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                4
On-line CPU(s) list:   0-3
Thread(s) per core:    2
Core(s) per socket:    2
Socket(s):             1
NUMA node(s):          1
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 37
Model name:            Intel(R) Core(TM) i5 CPU         660  @ 3.33GHz
Stepping:              2
CPU MHz:               3334.000
CPU max MHz:           3334.0000
CPU min MHz:           1200.0000
BogoMIPS:              6688.08
Virtualization:        VT-x
L1d cache:             32K
L1i cache:             32K
L2 cache:              256K
L3 cache:              4096K
NUMA node0 CPU(s):     0-3

**Comment by akopytov** [comment]
This is puzzling, actually. So all packages are intentionally built with the `--without-gcc-arch` configure flag. Which means no `-march` options are used by the compiler. Which is indeed the case, here's the build log for the 1.0.12 Jessie package: https://travis-ci.org/akopytov/sysbench/jobs/329809360

So what is the default `-march` value on GCC shipped in Jessie? It is apparently `x86-64`:
```
$ docker run -it debian:jessie
root@e55f6b5b1b33:/# apt update
...
root@e55f6b5b1b33:/# apt install -y gcc
...
root@e55f6b5b1b33:/# gcc --version
gcc (Debian 4.9.2-10) 4.9.2
Copyright (C) 2014 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
root@e55f6b5b1b33:/# gcc -Q --help=target | grep arch
  -march=                     		x86-64
```

As I understand, `--march=x86-64` **should** result in a binary which is compatible with Westmere. But something doesn't add up.

You mentioned you got a working binary after rebuilding the package. Which GCC version was used on your system? Can you repeat the above commands? Thanks.

```

**Comment by CHERTS** [comment]
This is info on the virtual server OpenVZ/Virtuozzo:

$ lsb_release -a
No LSB modules are available.
Distributor ID: Debian
Description:    Debian GNU/Linux 8.10 (jessie)
Release:        8.10
Codename:       jessie

$ gcc --version
gcc (Debian 4.9.2-10) 4.9.2

$ gcc -Q --help=target | grep arch
  -march=                               x86-64

**Comment by CHERTS** [comment]
I realized what the error was, your package has nothing to do with it.
I installed the Percona repository and the sysbench package was installed from it.
I did not notice this.

After
$ apt-get clean
and re-use
$ apt-get install sysbench
Reading package lists... Done
Building dependency tree
Reading state information... Done
The following NEW packages will be installed:
  sysbench
0 upgraded, 1 newly installed, 0 to remove and 2 not upgraded.
Need to get 326 kB of archives.
After this operation, 1,037 kB of additional disk space will be used.
Get:1 http://repo.percona.com/apt/ jessie/main sysbench amd64 1.0.12-1.jessie [326 kB]
Fetched 326 kB in 2s (161 kB/s)
Selecting previously unselected package sysbench.
(Reading database ... 38358 files and directories currently installed.)
Preparing to unpack .../sysbench_1.0.12-1.jessie_amd64.deb ...
Unpacking sysbench (1.0.12-1.jessie) ...
Setting up sysbench (1.0.12-1.jessie) ...

I saw where it was downloaded from.

I removed the Percona repository and installed the sysbench from packagecloud.io.
I'm sorry I took a lot of your time.
